src/wing/CCPACSWingCell.cpp

```diff
@@ -663,33 +815,139 @@ void CCPACSWingCell::BuildSkinGeometry(GeometryCache& cache) const
-    // cache points for later use
-    cache.IBLE = wsr.GetPoint(m_etaXsiCache->innerLeadingEdgePoint.eta,
-                              m_etaXsiCache->innerLeadingEdgePoint.xsi,
-                              WING_COORDINATE_SYSTEM);
-    cache.projectedIBLE = ProjectPointOnShape(loftShape, cache.IBLE, zRefDir);
-    cache.IBTE = wsr.GetPoint(m_etaXsiCache->innerTrailingEdgePoint.eta,
-                              m_etaXsiCache->innerTrailingEdgePoint.xsi,
-                              WING_COORDINATE_SYSTEM);
-    cache.projectedIBTE = ProjectPointOnShape(loftShape, cache.IBTE, zRefDir);
-    cache.OBLE = wsr.GetPoint(m_etaXsiCache->outerLeadingEdgePoint.eta,
-                              m_etaXsiCache->outerLeadingEdgePoint.xsi,
-                              WING_COORDINATE_SYSTEM);
-    cache.projectedOBLE = ProjectPointOnShape(loftShape, cache.OBLE, zRefDir);
-    cache.OBTE = wsr.GetPoint(m_etaXsiCache->outerTrailingEdgePoint.eta,
-                              m_etaXsiCache->outerTrailingEdgePoint.xsi,
-                              WING_COORDINATE_SYSTEM);
-    cache.projectedOBTE = ProjectPointOnShape(loftShape, cache.OBTE, zRefDir);
-
-    // cut the shape at the cell borders
-    TopoDS_Shape resultShape = CutSpanwise(cache, loftShape, SpanWiseBorder::Inner, m_positioningInnerBorder, zRefDir, 1e-2);
-    resultShape              = CutSpanwise(cache, resultShape, SpanWiseBorder::Outer, m_positioningOuterBorder, zRefDir, 1e-2);
-    resultShape              = CutChordwise(cache, resultShape, ChordWiseBorder::LE, m_positioningLeadingEdge, zRefDir, 1e-2);
-    resultShape              = CutChordwise(cache, resultShape, ChordWiseBorder::TE, m_positioningTrailingEdge, zRefDir, 1e-2);
-
-    // store the result
-    cache.skinGeometry = resultShape;
+    int use_contour_coordinates = (int)(m_positioningInnerBorder.GetInputType()  == CCPACSWingCellPositionSpanwise::Contour);
+    use_contour_coordinates    += (int)(m_positioningOuterBorder.GetInputType()  == CCPACSWingCellPositionSpanwise::Contour);
+    use_contour_coordinates    += (int)(m_positioningLeadingEdge.GetInputType()  == CCPACSWingCellPositionChordwise::Contour);
+    use_contour_coordinates    += (int)(m_positioningTrailingEdge.GetInputType() == CCPACSWingCellPositionChordwise::Contour);
+
+    if ( use_contour_coordinates > 0 && use_contour_coordinates != 4 ) {
+        throw CTiglError("If one boundary of a wing cell is defined using contour coordinates, "
+                         "all boundaries must be defined using contour coordinates");
+    }
+
+    if ( use_contour_coordinates ) {
+
+        // The skin is composed of a rectangular grid of faces. We need to annotate
+        // these faces to trim the correct faces at the correct parameters.
+
+        cache.rgsurface.SetShape(loftShape);
+
+        /*
+         *  Step 1/3:
+         *
+         *  create sums of all parameter ranges for every row (in spanwise v-dir)
```

Comment:
IMHO, this functionality should be better placed into the `CTiglRectGridSurface` class.